govcd/vm.go

```diff
@@ -1530,3 +1523,214 @@ func (vm *VM) UpdateVmCpuAndMemoryHotAddAsync(cpuHot, memoryAdd bool) (Task, err
+	task, err := addEmptyVmAsyncV10(vapp, reComposeVAppParams)
+	if err != nil {
+		return nil, err
+	}
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
+
+	err = vapp.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing vapp: %s", err)
+	}
+
+	//vApp Might Not Have Any VMs
+	if vapp.VApp.Children == nil {
+		return nil, ErrorEntityNotFound
+	}
+
+	util.Logger.Printf("[TRACE] Looking for VM: %s", reComposeVAppParams.CreateItem.Name)
+	for _, child := range vapp.VApp.Children.VM {
+
+		util.Logger.Printf("[TRACE] Looking at: %s", child.Name)
+		if child.Name == reComposeVAppParams.CreateItem.Name {
+			return getVMByHrefV10(vapp.client, child.HREF)
+		}
+
+	}
+	util.Logger.Printf("[TRACE] Couldn't find VM: %s", reComposeVAppParams.CreateItem.Name)
+	return nil, ErrorEntityNotFound
+}
+
+// addEmptyVmAsyncV97 adds an empty VM (without template) to the vApp and returns a Task and an error.
+func addEmptyVmAsyncV97(vapp *VApp, reComposeVAppParams *types.RecomposeVAppParamsForEmptyVm) (Task, error) {
+	err := validateEmptyVmParams(reComposeVAppParams)
+	if err != nil {
+		return Task{}, err
+	}
+	apiEndpoint, _ := url.ParseRequestURI(vapp.VApp.HREF)
+	apiEndpoint.Path += "/action/recomposeVApp"
+
+	reComposeVAppParams.XmlnsVcloud = types.XMLNamespaceVCloud
+	reComposeVAppParams.XmlnsOvf = types.XMLNamespaceOVF
+
+	// Return the task
+	return vapp.client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodPost,
+		types.MimeRecomposeVappParams, "error instantiating a new VM: %s", reComposeVAppParams)
+}
+
+// addEmptyVmV97 adds an empty VM (without template) to vApp and returns the new created VM or an error.
+func addEmptyVmV97(vapp *VApp, reComposeVAppParams *types.RecomposeVAppParamsForEmptyVm) (*VM, error) {
+	task, err := addEmptyVmAsyncV97(vapp, reComposeVAppParams)
+	if err != nil {
+		return nil, err
+	}
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
+
+	err = vapp.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing vapp: %s", err)
+	}
+
+	//vApp Might Not Have Any VMs
+	if vapp.VApp.Children == nil {
+		return nil, ErrorEntityNotFound
+	}
+
+	util.Logger.Printf("[TRACE] Looking for VM: %s", reComposeVAppParams.CreateItem.Name)
+	for _, child := range vapp.VApp.Children.VM {
+
+		util.Logger.Printf("[TRACE] Looking at: %s", child.Name)
+		if child.Name == reComposeVAppParams.CreateItem.Name {
+			return getVMByHrefV97(vapp.client, child.HREF)
+		}
+
+	}
+	util.Logger.Printf("[TRACE] Couldn't find VM: %s", reComposeVAppParams.CreateItem.Name)
+	return nil, ErrorEntityNotFound
+}
+
+// getVMByHrefV10 returns a VM reference by running a vCD API call
+// If no valid VM is found, it returns a nil VM reference and an error
+// Note that the pointer receiver here is a Client instead of a VApp, because
+// there are cases where we know the VM HREF but not which VApp it belongs to.
+func getVMByHrefV10(client *Client, vmHref string) (*VM, error) {
```

Comment:
Please add a comment defining `V10` (that it's referring to VCD version) and why do we need different functions for 10 and 9.7